File: ScapesCore/src/main/java/org/tobi29/scapes/packets/PacketInteraction.java
/*
 * Copyright 2012-2016 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.packets;

import java8.util.Optional;
import org.tobi29.scapes.chunk.WorldClient;
import org.tobi29.scapes.client.connection.ClientConnection;
import org.tobi29.scapes.engine.server.InvalidPacketDataException;
import org.tobi29.scapes.engine.utils.io.ReadableByteStream;
import org.tobi29.scapes.engine.utils.io.WritableByteStream;
import org.tobi29.scapes.entity.client.EntityClient;
import org.tobi29.scapes.entity.client.MobPlayerClient;
import org.tobi29.scapes.server.connection.PlayerConnection;

import java.io.IOException;

public class PacketInteraction extends PacketAbstract
        implements PacketClient, PacketServer {
    public static final byte INVENTORY_SLOT_LEFT = 0x00, INVENTORY_SLOT_RIGHT =
            0x01, OPEN_INVENTORY = 0x10, CLOSE_INVENTORY = 0x11;
    private int entityID;
    private byte type, data;

    public PacketInteraction() {
    }

    public PacketInteraction(byte type) {
        this(type, (byte) 0);
    }

    public PacketInteraction(byte type, byte data) {
        this.type = type;
        this.data = data;
    }

    public PacketInteraction(int entityID, byte type, byte data) {
        this.entityID = entityID;
        this.type = type;
        this.data = data;
    }

    @Override
    public void sendClient(PlayerConnection player, WritableByteStream stream)
            throws IOException {
        stream.putInt(entityID);
        stream.put(type);
        stream.put(data);
    }

    @Override
    public void parseClient(ClientConnection client, ReadableByteStream stream)
            throws IOException {
        entityID = stream.getInt();
        type = stream.get();
        data = stream.get();
    }

    @Override
    public void runClient(ClientConnection client, WorldClient world) {
        if (world == null) {
            return;
        }
        Optional<EntityClient> fetch = world.entity(entityID);
        if (fetch.isPresent()) {
            EntityClient entity = fetch.get();
            if (entity instanceof MobPlayerClient) {
                switch (type) {
                    case INVENTORY_SLOT_LEFT:
                        ((MobPlayerClient) entity).setInventorySelectLeft(data);
                        break;
                    case INVENTORY_SLOT_RIGHT:
                        ((MobPlayerClient) entity)
                                .setInventorySelectRight(data);
                        break;
                }
            }
        } else {
            client.send(new PacketRequestEntity(entityID));
        }
    }

    @Override
    public void sendServer(ClientConnection client, WritableByteStream stream)
            throws IOException {
        stream.put(type);
        stream.put(data);
    }

    @Override
    public void parseServer(PlayerConnection player, ReadableByteStream stream)
            throws IOException {
        type = stream.get();
        data = stream.get();
    }

    @Override
    public void runServer(PlayerConnection player) {
        player.mob(mob -> {
            switch (type) {
                case INVENTORY_SLOT_LEFT:
                    if (data < 0 || data >= 10) {
                        throw new InvalidPacketDataException(
                                "Invalid slot change data!");
                    }
                    mob.setInventorySelectLeft(data);
                    mob.world().send(new PacketInteraction(mob.entityID(),
                            INVENTORY_SLOT_LEFT, data));
                    break;
                case INVENTORY_SLOT_RIGHT:
                    if (data < 0 || data >= 10) {
                        throw new InvalidPacketDataException(
                                "Invalid slot change data!");
                    }
                    mob.setInventorySelectRight(data);
                    mob.world().send(new PacketInteraction(mob.entityID(),
                            INVENTORY_SLOT_RIGHT, data));
                    break;
                case OPEN_INVENTORY:
                    player.send(new PacketOpenGui(mob));
                    break;
                case CLOSE_INVENTORY:
                    mob.inventories().modify("Hold",
                            inventory -> inventory.item(0).take()
                                    .ifPresent(mob::dropItem));
                    player.send(new PacketCloseGui());
                    break;
                default:
                    throw new InvalidPacketDataException(
                            "Invalid interaction type!");
            }
        });
    }
}

File: ScapesCore/src/main/kotlin/org/tobi29/scapes/packets/PacketInteraction.kt
/*
 * Copyright 2012-2017 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.tobi29.scapes.packets

import org.tobi29.scapes.block.Registries
import org.tobi29.scapes.client.connection.ClientConnection
import org.tobi29.scapes.engine.server.InvalidPacketDataException
import org.tobi29.scapes.engine.utils.io.ReadableByteStream
import org.tobi29.scapes.engine.utils.io.WritableByteStream
import org.tobi29.scapes.entity.client.MobPlayerClient
import org.tobi29.scapes.server.connection.PlayerConnection
import java.util.*

class PacketInteraction : PacketAbstract, PacketBoth {
    private lateinit var uuid: UUID
    private var side: Byte = 0
    private var data: Byte = 0

    constructor(type: PacketType) : super(type)

    constructor(type: PacketType,
                side: Byte,
                data: Byte) : super(type) {
        this.side = side
        this.data = data
    }

    constructor(type: PacketType,
                uuid: UUID,
                side: Byte,
                data: Byte) : super(type) {
        this.uuid = uuid
        this.side = side
        this.data = data
    }

    constructor(registry: Registries,
                side: Byte,
                data: Byte) : this(
            Packet.make(registry, "core.packet.Interaction"), side, data)

    constructor(registry: Registries,
                uuid: UUID,
                side: Byte,
                data: Byte) : this(
            Packet.make(registry, "core.packet.Interaction"), uuid, side,
            data)

    override fun sendClient(player: PlayerConnection,
                            stream: WritableByteStream) {
        stream.putLong(uuid.mostSignificantBits)
        stream.putLong(uuid.leastSignificantBits)
        stream.put(side.toInt())
        stream.put(data.toInt())
    }

    override fun parseClient(client: ClientConnection,
                             stream: ReadableByteStream) {
        uuid = UUID(stream.long, stream.long)
        side = stream.get()
        data = stream.get()
    }

    override fun runClient(client: ClientConnection) {
        client.getEntity(uuid) { entity ->
            if (entity is MobPlayerClient) {
                when (side) {
                    INVENTORY_SLOT_LEFT -> entity.inventorySelectLeft = data.toInt()
                    INVENTORY_SLOT_RIGHT -> entity.inventorySelectRight = data.toInt()
                }
            }
        }
    }

    override fun sendServer(client: ClientConnection,
                            stream: WritableByteStream) {
        stream.put(side.toInt())
        stream.put(data.toInt())
    }

    override fun parseServer(player: PlayerConnection,
                             stream: ReadableByteStream) {
        side = stream.get()
        data = stream.get()
    }

    override fun runServer(player: PlayerConnection) {
        player.mob { mob ->
            when (side) {
                INVENTORY_SLOT_LEFT -> {
                    if (data < 0 || data >= 10) {
                        throw InvalidPacketDataException(
                                "Invalid slot change data!")
                    }
                    mob.inventorySelectLeft = data.toInt()
                    mob.world.send(PacketInteraction(type, mob.getUUID(),
                            INVENTORY_SLOT_LEFT, data))
                }
                INVENTORY_SLOT_RIGHT -> {
                    if (data < 0 || data >= 10) {
                        throw InvalidPacketDataException(
                                "Invalid slot change data!")
                    }
                    mob.inventorySelectRight = data.toInt()
                    mob.world.send(PacketInteraction(type, mob.getUUID(),
                            INVENTORY_SLOT_RIGHT, data))
                }
                OPEN_INVENTORY -> {
                    mob.onOpenInventory()
                    player.send(
                            PacketOpenGui(player.server.plugins.registry, mob))
                }
                CLOSE_INVENTORY -> {
                    mob.onCloseInventory()
                    player.send(
                            PacketCloseGui(player.server.plugins.registry))
                }
                else -> throw InvalidPacketDataException(
                        "Invalid interaction type!")
            }
        }
    }

    companion object {
        val INVENTORY_SLOT_LEFT: Byte = 0x00
        val INVENTORY_SLOT_RIGHT: Byte = 0x01
        val OPEN_INVENTORY: Byte = 0x10
        val CLOSE_INVENTORY: Byte = 0x11
    }
}
